lib/plugins/aws/deploy/compile/functions/index.js

```diff
@@ -14,54 +14,86 @@ class AwsCompileFunctions {
     };
   }
 
+  /**
+   * Compile the declared functions into the Cloud Formation Template (CFT)
+   */
   compileFunctions() {
+    if (this.anyFunctionHasNoRole()) {
+      this.attachDefaultRoleAndPolicy();
+    }
+    this.mergeFunctionsIntoCft();
+  }
+
+  /**
+   * Check whether any function is missing an assigned role.
+   * All functions have a role if a provider-wide iamRoleARN is declared.  If one is not declared
+   * then each function must declare its own iamRoleARN.  Each function may override a declared
+   * provider level iamRoleARN by declaring its own.
+   * @returns {boolean} Whether any of the declared functions did not have an iamRoleARN.
+   */
+  anyFunctionHasNoRole() {
+    let ret = false;
     if (typeof this.serverless.service.provider.iamRoleARN !== 'string') {
```

Comment:
I believe there is a bug in this.  if iamRoleArn is an object (i.e. `{ 'Fn::GetAtt': ['IamRoleLambdaExecution', 'Arn'] }`) then this doesn't detect the declaration of that  arn declaration.